internal/resources/metal/connection/datasource_schema.go

```diff
@@ -0,0 +1,158 @@
+			Required:    true,
+		},
+		"name": schema.StringAttribute{
+			Description: "Name of the connection resource",
+			Computed:    true,
+		},
+		"facility": schema.StringAttribute{
+			Description:        "Facility which the connection is scoped to",
+			Computed:           true,
+			DeprecationMessage: "Use metro instead of facility. For more information, read the migration guide.",
+		},
+		"metro": schema.StringAttribute{
+			Description: "Metro which the connection is scoped to",
+			Computed:    true,
+		},
+		"redundancy": schema.StringAttribute{
+			Description: "Connection redundancy - redundant or primary",
+			Computed:    true,
+		},
+		"contact_email": schema.StringAttribute{
+			Description: "The preferred email used for communication and notifications about the Equinix Fabric interconnection",
+			Computed:    true,
+		},
+		"type": schema.StringAttribute{
+			Description: "Connection type - dedicated or shared",
+			Computed:    true,
+		},
+		"project_id": schema.StringAttribute{
+			Description: "ID of project to which the connection belongs",
+			Computed:    true,
+		},
+		"speed": schema.StringAttribute{
+			Description: fmt.Sprintf("Port speed. Possible values are %s", allowedSpeedsString()),
+			Computed:    true,
+		},
+		"description": schema.StringAttribute{
+			Description: "Description of the connection resource",
+			Computed:    true,
+		},
+		"mode": schema.StringAttribute{
+			Description: fmt.Sprintf("Connection mode - %s or %s",
+				string(packngo.ConnectionModeStandard),
+				string(packngo.ConnectionModeTunnel),
+			),
+			Computed: true,
+		},
+		"tags": schema.ListAttribute{
+			Description: "Tags attached to the connection",
+			Computed:    true,
+			ElementType: types.StringType,
+		},
+		"vlans": schema.ListAttribute{
+			Description: "Attached vlans, only in shared connection",
+			Computed:    true,
+			ElementType: types.Int64Type,
+		},
+		"service_token_type": schema.StringAttribute{
+			Description: "Only used with shared connection. Type of service token to use for the connection, a_side or z_side",
+			Computed:    true,
+		},
+		"organization_id": schema.StringAttribute{
+			Description: "ID of organization to which the connection is scoped to",
+			Computed:    true,
+		},
+		"status": schema.StringAttribute{
+			Description: "Status of the connection resource",
+			Computed:    true,
+		},
+		"token": schema.StringAttribute{
+			Description:        "Only used with shared connection. Fabric Token required to continue the setup process with [equinix_ecx_l2_connection](https://registry.terraform.io/providers/equinix/equinix/latest/docs/resources/equinix_ecx_l2_connection) or from the [Equinix Fabric Portal](https://ecxfabric.equinix.com/dashboard)",
+			Computed:           true,
+			DeprecationMessage: "If your organization already has connection service tokens enabled, use `service_tokens` instead",
+		},
+		"service_tokens": schema.ListNestedAttribute{
```

Comment:
It looks like `ListNestedAttributes` may be a protocol version 6 thing, and we're currently using protocol version 5: https://developer.hashicorp.com/terraform/plugin/framework/migrating/attributes-blocks/blocks-computed#framework

Reply:
Grasping at straws a bit here, because other than that migration page and the error message we're getting, I can't find anything that explicitly says that this function only works for protocol v6.  I pushed up a duplicate branch that includes a fix for the virtual circuit acceptance tests and converts the provider to protocol v6 to see if that resolves the issue: https://github.com/equinix/terraform-provider-equinix/actions/runs/7935657886/job/21669301881

If it does, though, we'll have to think about how to proceed, since switching to protocol v6 means dropping support for terraform < 1.0.  (From what I can tell, the protocol v5 version of this change would mean losing descriptions for the nested attributes.)

Reply:
Closing the loop here:  once everything was cut over to protocol v6 there were errors about types not supporting optional values, so probably the same problem that was actually happening here, except the error message with v6 was better.  In that case the only reason to upgrade would be to maintain our sanity, so:  no need to upgrade.